### ISSUE TITLE
MeshBuilder.CreateSphere creates a sphere with some triangles having lengths of 0, causing SubMesh's projectToRef to return invalid results 

### ISSUE THREAD

**Issue by rickfromwork** [BabylonJS/Babylon.js#10559]
### This breaks XR near interaction accuracy on scenes with spherical shapes, since we use this logic to determine the touched object

When creating a default sphere using this call, the first 37 indices have identical positions, at the top of the sphere. If this is iterated through via SubMesh's projectToMesh function, the normals/lengths inside math.vector's ProjectOnTriangleToRef get messed up.
The minimum length is capped at an epsilon of 0.001, so it takes the distance between two vertices of each affected triangle, 0, and sets it to 0.001. This and the vector that created that length - {0, 0, 0} - interferes with later calculations, leading to a zeroed-out normal vector that makes the function think the projected point is inside the triangle. 

This severely impacts where the point returned from projectToMesh appears, as shown in this example:
https://www.babylonjs-playground.com/#DWPQ9R#452

**Comment by rickfromwork** [comment]
There may be more or less identical vertices, depending on how many faces the sphere has. From my understanding when making the sphere it creates strips of alternating triangles, then wraps it around the sphere. At the top and bottom of the sphere though, the outer side of the strip is wrapped around a single point, which is why the positions are all identical.

I imagine a more efficient winding would only have half the triangles on the very top and bottom 'strips' of spherical shapes, since currently the other half of the triangles are all some permutation of {point A, point A, point B}.

**Comment by sebavan** [comment]
Adding @RaananW our in house XR master :-)

**Comment by sebavan** [comment]
@rickfromwork I wonder if in your case you could rely on an icoSphere which should have a more evenly distributed topology ?

**Comment by sebavan** [comment]
Yup it seems at least it could unlock you in the mean time ? https://www.babylonjs-playground.com/#DWPQ9R#453

**Comment by RaananW** [comment]
> @rickfromwork I wonder if in your case you could rely on an icoSphere which should have a more evenly distributed topology ?

icoSphere will probably be the right way to go

**Comment by rickfromwork** [comment]
Using an icosphere will not fix the issue that the webXRNearInteraction code hits, since the sphere/target mesh is a mesh in the scene, but it can unblock my testing if I make all the spheres in my test scenes for near interaction icospheres.

WebXRNearInteraction runs this logic seen in the playground scene on all meshes in the scene that are set to react to near interactions, where the mesh in the scene takes the place of sphereMain/boxMain, and the fingertip takes the place of sphereOuter. It wouldn't make sense to force developers to only enable near interaction on a subset of mesh types, but I can do that for now in my scenes.

**Comment by rickfromwork** [comment]
That being said, I didn't realize icospheres could look that _spherical_ until I saw the updated playground scene. I always thought of icospheres as low-poly spheres.

**Comment by sebavan** [comment]
So I guess we currently have 2 issues to solve:

[ ] CreateSphere should not create triangles of 0 length.
[ ]  ProjectOnTriangleToRef should work despite 0 length triangles.

Does this sound correct ?

**Comment by sebavan** [comment]
As this is not related to XR I ll be able to help on this one

**Comment by rickfromwork** [comment]
> So I guess we currently have 2 issues to solve:
> 
> * [x]  CreateSphere should not create triangles of 0 length.
> * [ ]  ProjectOnTriangleToRef should work despite 0 length triangles.
> 
> Does this sound correct ?

Yeah, that sounds good, thanks.

**Comment by sebavan** [comment]
So we do not have at the moment full 0 lenght triangle -> or dots in the sphere generation but the ProjectOnTriangleToRef is still having issues.

First there should be a special case for all length = to 0

```
// Special case if all vertices are in the same position
        if (p1p0L == 0 && p2p0L == 0 && p2p1L == 0) {
            ref.copyFrom(p0);
            vector.subtractToRef(p0, vectorp0);
            return vectorp0.length();
        }
```

Also we should then check triangles defined as lines (also producing a 0 normal)

Example for p1 and p0 being on a line:

```
// Special case if all vertices lie on a line
        if (p1p0L == 0) {
            vector.subtractToRef(p0, vectorp0);
            const unitVector = MathTmp.Vector3[5];
            p2p0.normalizeToRef(unitVector);
            const cosAvectorp0 = Vector3.Dot(vectorp0, unitVector);

            const projVector = MathTmp.Vector3[6];
            unitVector.scaleToRef(cosAvectorp0, projVector);

            const projPoint = MathTmp.Vector3[7];
            p0.addToRef(projVector, projPoint);

            const p0pp = MathTmp.Vector3[8];
            const p2pp = MathTmp.Vector3[9];
            projPoint.subtractToRef(p0, p0pp);
            projPoint.subtractToRef(p2, p2pp);

            // Let s move on the line with origin p0.
            const p0X = Vector3.Dot(p0pp.normalizeToNew(), unitVector);
            const p2X = Vector3.Dot(p2pp.normalizeToNew(), unitVector);

            if (p0X * p2X < 0) {
                ref.copyFrom(projPoint);
            }
            else if (p0pp.length() < p2pp.length()) {
                ref.copyFrom(p0);
            }
            else {
                ref.copyFrom(p2);
            }

            return Vector3.Distance(vector, ref);
        }
```

And after that there are still something wrong in the algorythm. The detection of projected point within a triangle is not working correctly. You can try with those inputs: 

{ x: 0, y: 1, z: 0 } { x: -0.06818714150362837, y: 0.9957341551780701, z: -0.06216079213354764 } { x: -0.06216079213354764, y: 0.9957341551780701, z: 0.06818714150362837 } and the vector { x: 1.5, y: 0.5, z: 1.5 } (coming from the previous playground)

@CraigFeldspar could you check to be sure your section (// Determines if "proj" lies inside the triangle) is correct ?

**Comment by sebavan** [comment]
Actually @CraigFeldspar is in vacation so I might pull the brain from @CedricGuillemet for this one :-)

**Comment by sebavan** [comment]
@CraigFeldspar will have a look on Tuesday :-)

**Comment by CraigFeldspar** [comment]
I just linked a to a PR to handle degenerate triangles, which is not correct regarding maths, but is good enough if we assume no mesh is gonna be fully degenerate. But I just realized @sebavan already put code in this issue to handle these cases better, so feel free to replace with this better code in the PR.

Regarding the algorithm problem that you mentionned @sebavan, I find myself having correct results with this playground (on my branch), reproducing your case : http://localhost:1338/Playground/index-local.html#YTYQRN

`proj` is :
```
x: 1.4614052295378541
y: 1.0909147977169313
z: 1.5017843441332208
``` 
which is a correct projection of  `x: 1.5, y: 0.5, z: 1.5` on the quasi y-aligned triangle that you specified as an input. Then the result is clamped to `x: -0.06216079213354764, _y: 0.9957341551780701, _z: 0.06818714150362837`.

I also drew a PG that displays the projection of the intersection of the mouse & slate with the top bar (with a max radius of 0.3), and everything seems to work ok, at least in the case of no degenerate triangles : http://localhost:1338/Playground/index-local.html#2YZFA0#239

Let me know if you find another non-working example
